139 Initial TS Migration, Setup TS Jest, and Update ESLint

hiyaryan opened the pull request:
This PR sets up the initial TS Jest testing environment and uses mongodb-memory-server. The following changes were made (also in commit body) in order to allow Jest to understand TypeScript. A test has been made for `index.ts` that itself tests this setup. Additionally, some minor changes have been added, adding/updating comments in the files worked on and including a graceful shutdown in `index.ts`.

- Updated `package.json`:
  - Changed the development script to use `tsc --watch` instead of `tsx watch`.
  - Added development dependencies: `@babel/preset-typescript`, `@types/express`, `@types/jest`, `mongodb-memory-server`, and `ts-node`.

- Modified `db.js`:
  - Replaced default import with named import for `mongoose`.
  - Exported `mongoose` as `db` to ensure consistent type definitions.

- Updated `index.ts`:
  - Added comments and structured the server startup logic.
  - Implemented a graceful shutdown mechanism using `process.on` to listen for SIGINT and SIGTERM signals.

- Created `index.test.ts`:
  - Added initial Jest tests to verify server startup in non-production mode.
  - Mocked console.log and added assertions to check server behavior.

- Modified Jest setup and teardown scripts:
  - Simplified the setup by combining imports and logging the seeding process.
  - Added a console log statement in the teardown script for clarity.

davidzlu commented on backend/data/seed.js:
```diff
@@ -82,22 +92,22 @@ export async function seedDatabase() {
     console.error('Error while seeding database:', error);
   } finally {
-    await mongoose.disconnect();
+    !memoryMongo && await mongoose.disconnect();
```
To clarify, this line is meant to be like a compact if statement? The goal is to have mongoose disconnect only if it's something other than a MongoMemoryServer in use?

hiyaryan replied:
This is called a [short-circuit evaluation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_AND#short-circuit_evaluation). Essentially, if the left side is false, since logical AND requires both sides to be true, it simply stops and doesn't evaluate the right side of &&.

In this case, if we already have mongod running in the background either locally or in a container, we're just using `mongoose` to help us seed the mongo database then we disconnect properly. But when it comes to using the `memoryMongo`, we don't want to close it because it's keeping the MongoMemoryServer alive. If we disconnect it the database will close and that will cause our tests that use it to have a connection error.

davidzlu replied:
Gotcha thanks for the explanation. I think it comes down to taste and what you think the code standards are, but I'm not a fan of checking a conditional like this. It makes the branching implicit, and I haven't seen this style used anywhere else in the codebase. I'd suggest doing a single-line if statement instead, like `if (!memoryMongo) await mongoose.disconnect();`. But if you feel strongly about doing it this way I won't stop you.

hiyaryan replied:
I agree. It's more readable too using a single-line if statement. You use short-circuiting often in JSX. This is the only place it's used and it's not that easy to understand so I'll update it!

davidzlu commented on backend/src/index.ts:
```diff
@@ -1,28 +1,66 @@
+  // Start the server in production mode
+  const { PRIVATE_KEY_PATH, CERTIFICATE_PATH, CA_PATH } = process.env;
+
+  if (!PRIVATE_KEY_PATH || !CERTIFICATE_PATH || !CA_PATH) {
```
I appreciate the explicit configuration checks and clear error message! Very helpful

davidzlu commented on backend/src/db.js:
```diff
@@ -26,3 +26,5 @@ export default async function connectDB(dbName) {
   }
 }
+
+export { mongoose as db };
```
I'm not sure if I understand why this line is needed? It doesn't seem like anything happens with mongoose, and other classes could just import mongoose directly rather than through this module. Could you please explain what the intent behind this is?

hiyaryan replied:
Think of it like a Singleton. If we export the same instance of mongoose we can run operations against it anywhere in the code base. Mongoose helps establish a connection to a Mongo database and close the connection when we don't need it. In the index.test.ts file, after running the test, Jest said that it detected an open handle and it didn't exit, this is because the MongoDB connection was still open. The Mongoose instance I exported from db.js lets us close that connection and properly disconnect from the DB.

davidzlu replied:
I see so you're trying to cache the mongoose connection right? I'm not sure if it's needed given that when you call mongoose.connect(), mongoose creates a default connection that you access with mongoose.connection (https://mongoosejs.com/docs/connections.html#multiple_connections). What was the exact error you were getting? The only kind of connection error I get is when the Assistant.js test suite runs, and I kinda expect that because of things are setup on my computer right now.

I see that one way of caching a mongoose connection the docs recommends is to store the connection promise in a variable and check it before calling mongoose.connect() again (https://mongoosejs.com/docs/lambda.html). Whether or not you want to do it this way seems to depend on how you deploy the app. I'd think about this later in a separate issue.

hiyaryan replied:
So in other words mongoose itself is a singleton. Importing mongoose from mongoose as opposed to from `db.js` should close the connection. And I tested it, and it does! The issue I was facing was that the test wasn't exiting cleanly because of an [open handle](https://jestjs.io/docs/cli#--detectopenhandles) due mongoose not being explicitly closed. But importing mongoose itself made it work! I'll revert this file.

davidzlu reviewed (changes requested):
Overall it's looking really good! Great job! The main thing is I want to make sure we understand what the babel typescript preset does and doesn't do, and confirm that it fits our needs. The rest of the review is a couple of questions I had and edits I think would be helpful.

hiyaryan commented on backend/package.json:
```diff
@@ -5,7 +5,7 @@
   "scripts": {
-    "dev": "npm install && tsx watch src/index.ts",
+    "dev": "npm install && tsc --watch",
     "test": "NODE_ENV=test jest --coverage --detectOpenHandles",
```
Awesome! Will remove. Hopefully that fixes your issue with running the command.

As a side note, if you run `npm run test` all the old tests will fail (or most will). You probably already noticed that so be sure to isolate only the tests we're working on to prevent any unforeseen issues.

hiyaryan commented on backend/data/seed.js:
```diff
@@ -26,9 +29,16 @@ async function seedConfigs() {
+      console.log("Setting up in-memory database...")
+      await memoryMongo.start();
+      mongoURI = memoryMongo.getUri();
+     }
```
That's embarrassing 😅 Nice catch! I kept telling myself I'd set up the eslint rules so that saving in VS Code will automatically apply them and lint the code. It's been a while since I worked on the backend and forgot to run the linter manually.

hiyaryan commented:
> Overall it's looking really good! Great job! The main thing is I want to make sure we understand what the babel typescript preset does and doesn't do, and confirm that it fits our needs. The rest of the review is a couple of questions I had and edits I think would be helpful.

Awesome! Thank you for the reviews @davidzlu. I should have resolved and responded to all of them. Most importantly, I added type checking to the tests and removed that pesky extra whitespace. Please have a look and let me know what you think!

davidzlu commented on backend/data/seed.js:
```diff
@@ -26,9 +29,16 @@ async function seedConfigs() {
+
+    // If testing, use the in-memory database
+    if (memoryMongo) { 
+      console.log("Setting up in-memory database...")
```
Nitpick: this line is missing a semicolon. Sorry for missing it 🤣

hiyaryan replied:
Alright, I'm sure there's a lot more of these kinds of problems. I'm going to update the eslintrc.cjs file with some additional rules, set up the linter in my IDE then fix them all in one go. I was considering doing it in a separate PR but I think the sooner the better so I stop making these kinds or mistakes 😆

hiyaryan replied:
@davidzlu Backend should be nice and pretty now 🙂

davidzlu commented on backend/package.json:
```diff
@@ -5,8 +5,8 @@
-    "dev": "npm install && tsx watch src/index.ts",
-    "test": "NODE_ENV=test jest --coverage --detectOpenHandles",
+    "dev": "npm install && tsc --watch",
+    "test": "tsc --project tsconfig.test.json && jest --coverage --detectOpenHandles",
```
The type checking works great! Nice job

davidzlu reviewed (approved):
Things are looking good! I've got a couple of suggestions and a style nitpick, but I'm find merging the changes as is.

davidzlu reviewed (approved):
Looks like you addressed all of my previous suggestions! I just checked your second to last commit and the .eslintrc.cjs file, but things seem good.
